pgmpy/causal_discovery/PC.py

```diff
@@ -0,0 +1,261 @@
+
+        if n_features == 0:
+            raise ValueError(
+                f"0 feature(s) (shape={X.shape}) while a minimum of 1 is required."
+            )
+        if n_samples < 2:
+            raise ValueError(f"n_samples = {n_samples}, at least 2 are required.")
+
+        # Handle cases like complex data, sparse arrays etc. first
+        if isinstance(X, pd.DataFrame):
+            _nodes = X.columns
+        else:
+            _nodes = [f"x{i}" for i in range(X.shape[1])]
+
+        X = validate_data(
+            self,
+            X=X,
+            dtype="numeric",
+            accept_sparse=False,
+            ensure_all_finite=True,
+            reset=True,  # reset=True in fit, reset=False in predict/transform
+        )
+
+        X = X.astype(float, copy=False)
+        X = pd.DataFrame(X, columns=_nodes)
+
+        # CI test
+        ci_test = get_callable_ci_test(self.ci_test, data=X)
+
+        if self.expert_knowledge is None:
+            expert_knowledge = ExpertKnowledge()
+        else:
+            expert_knowledge = self.expert_knowledge
+
+        if expert_knowledge.search_space:
+            expert_knowledge.limit_search_space(X.columns)
+
+        # Step 1: skeleton
+        skel, separating_sets = self.build_skeleton(
+            data=X,
+            independencies=independencies,
+            variant=self.variant,
+            ci_test=ci_test,
+            significance_level=self.significance_level,
+            max_cond_vars=self.max_cond_vars,
+            expert_knowledge=expert_knowledge,
+            enforce_expert_knowledge=self.enforce_expert_knowledge,
+            n_jobs=self.n_jobs,
+            show_progress=self.show_progress,
+        )
+
+        if self.return_type == "skeleton":
+            self.graph_ = (skel, separating_sets)
+            return self
+
+        # Step 2: orient colliders
+        pdag = self.orient_colliders(
+            skel, separating_sets, expert_knowledge.temporal_ordering
+        )
+
+        # Step 3: apply rules / expert knowledge
+        if expert_knowledge.temporal_order != [[]]:
+            pdag = expert_knowledge.apply_expert_knowledge(pdag)
+            pdag = pdag.apply_meeks_rules(apply_r4=True)
+        elif not self.enforce_expert_knowledge:
+            pdag = pdag.apply_meeks_rules(apply_r4=False)
+            pdag = expert_knowledge.apply_expert_knowledge(pdag)
+            pdag = pdag.apply_meeks_rules(apply_r4=True)
+        else:
+            pdag = pdag.apply_meeks_rules(apply_r4=False)
+
+        pdag.add_nodes_from(set(X.columns) - set(pdag.nodes()))
+
+        if self.return_type in ("pdag", "cpdag"):
+            self.graph_ = pdag
+        elif self.return_type == "dag":
+            self.graph_ = pdag.to_dag()
+        else:
+            raise ValueError(
+                f"return_type must be one of: dag, pdag, cpdag, skeleton. Got: {self.return_type}"
+            )
+
+        return self
+
+    @staticmethod
+    def orient_colliders(

```

Comment:
I think we should rename this to `_orient_colliders` to make it a private method.